File: trunk/symmetric-core/src/main/java/org/jumpmind/symmetric/service/impl/PushService.java
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.jumpmind.symmetric.service.impl;

import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.commons.lang.StringUtils;
import org.jumpmind.symmetric.common.ParameterConstants;
import org.jumpmind.symmetric.db.ISymmetricDialect;
import org.jumpmind.symmetric.model.BatchAck;
import org.jumpmind.symmetric.model.Node;
import org.jumpmind.symmetric.model.NodeCommunication;
import org.jumpmind.symmetric.model.NodeCommunication.CommunicationType;
import org.jumpmind.symmetric.model.NodeSecurity;
import org.jumpmind.symmetric.model.OutgoingBatch;
import org.jumpmind.symmetric.model.ProcessInfo;
import org.jumpmind.symmetric.model.ProcessInfo.Status;
import org.jumpmind.symmetric.model.ProcessInfoKey;
import org.jumpmind.symmetric.model.ProcessInfoKey.ProcessType;
import org.jumpmind.symmetric.model.RemoteNodeStatus;
import org.jumpmind.symmetric.model.RemoteNodeStatuses;
import org.jumpmind.symmetric.service.ClusterConstants;
import org.jumpmind.symmetric.service.IAcknowledgeService;
import org.jumpmind.symmetric.service.IClusterService;
import org.jumpmind.symmetric.service.IConfigurationService;
import org.jumpmind.symmetric.service.IDataExtractorService;
import org.jumpmind.symmetric.service.INodeCommunicationService;
import org.jumpmind.symmetric.service.INodeCommunicationService.INodeCommunicationExecutor;
import org.jumpmind.symmetric.service.INodeService;
import org.jumpmind.symmetric.service.IParameterService;
import org.jumpmind.symmetric.service.IPushService;
import org.jumpmind.symmetric.statistic.IStatisticManager;
import org.jumpmind.symmetric.transport.IOutgoingWithResponseTransport;
import org.jumpmind.symmetric.transport.ITransportManager;

/**
 * @see IPushService
 */
public class PushService extends AbstractOfflineDetectorService implements IPushService,
        INodeCommunicationExecutor {

    private IDataExtractorService dataExtractorService;

    private IAcknowledgeService acknowledgeService;

    private ITransportManager transportManager;

    private INodeService nodeService;

    private IClusterService clusterService;

    private INodeCommunicationService nodeCommunicationService;
    
    private IStatisticManager statisticManager;
    
    private IConfigurationService configurationService;

    private Map<String, Date> startTimesOfNodesBeingPushedTo = new HashMap<String, Date>();

    public PushService(IParameterService parameterService, ISymmetricDialect symmetricDialect,
            IDataExtractorService dataExtractorService, IAcknowledgeService acknowledgeService,
            ITransportManager transportManager, INodeService nodeService,
            IClusterService clusterService, INodeCommunicationService nodeCommunicationService, IStatisticManager statisticManager, IConfigurationService configrationService) {
        super(parameterService, symmetricDialect);
        this.dataExtractorService = dataExtractorService;
        this.acknowledgeService = acknowledgeService;
        this.transportManager = transportManager;
        this.nodeService = nodeService;
        this.clusterService = clusterService;
        this.nodeCommunicationService = nodeCommunicationService;
        this.statisticManager = statisticManager;
        this.configurationService = configrationService;
    }

    public Map<String, Date> getStartTimesOfNodesBeingPushedTo() {
        return new HashMap<String, Date>(startTimesOfNodesBeingPushedTo);
    }

    synchronized public RemoteNodeStatuses pushData(boolean force) {
        RemoteNodeStatuses statuses = new RemoteNodeStatuses(configurationService.getChannels(false));
        
        Node identity = nodeService.findIdentity(false);
        if (identity != null && identity.isSyncEnabled()) {
            long minimumPeriodMs = parameterService.getLong(ParameterConstants.PUSH_MINIMUM_PERIOD_MS, -1);
            if (force || !clusterService.isInfiniteLocked(ClusterConstants.PUSH)) {
                    List<NodeCommunication> nodes = nodeCommunicationService
                            .list(CommunicationType.PUSH);
                    if (nodes.size() > 0) {
                        NodeSecurity identitySecurity = nodeService.findNodeSecurity(identity
                                .getNodeId());
                        if (identitySecurity != null) {
                            int availableThreads = nodeCommunicationService
                                    .getAvailableThreads(CommunicationType.PUSH);
                            for (NodeCommunication nodeCommunication : nodes) {
                                boolean meetsMinimumTime = true;
                                if (minimumPeriodMs > 0 && nodeCommunication.getLastLockTime() != null &&
                                   (System.currentTimeMillis() - nodeCommunication.getLastLockTime().getTime()) < minimumPeriodMs) {
                                   meetsMinimumTime = false; 
                                }
                                if (availableThreads > 0 && meetsMinimumTime) {
                                    if (nodeCommunicationService.execute(nodeCommunication, statuses,
                                            this)) {
                                        availableThreads--;
                                    }
                                }
                            }
                        } else {
                            log.error(
                                    "Could not find a node security row for '{}'.  A node needs a matching security row in both the local and remote nodes if it is going to authenticate to push data",
                                    identity.getNodeId());
                        }
                    }
            } else {
                log.debug("Did not run the push process because it has been stopped");
            }
        }
        return statuses;
    }

    public void execute(NodeCommunication nodeCommunication, RemoteNodeStatus status) {
        Node node = nodeCommunication.getNode();
        if (StringUtils.isNotBlank(node.getSyncUrl()) || 
                !parameterService.isRegistrationServer()) {
            try {
                startTimesOfNodesBeingPushedTo.put(node.getNodeId(), new Date());
                long reloadBatchesProcessed = 0;
                long lastBatchCount = 0;
                do {
                    if (lastBatchCount > 0) {
                        log.info(
                                "Pushing to {} again because the last push contained reload batches",
                                node);
                    }
                    reloadBatchesProcessed = status.getReloadBatchesProcessed();
                    log.debug("Push requested for {}", node);
                    pushToNode(node, status);
                    if (!status.failed() && status.getBatchesProcessed() > 0 
                            && status.getBatchesProcessed() != lastBatchCount) {
                        log.info(
                                "Pushed data to {}. {} data and {} batches were processed",
                                new Object[] { node, status.getDataProcessed(),
                                        status.getBatchesProcessed()});
                    } else if (status.failed()) {
                        log.warn(
                                "There was a failure while pushing data to {}. {} data and {} batches were processed",
                                new Object[] { node, status.getDataProcessed(),
                                        status.getBatchesProcessed()});                        
                    }
                    log.debug("Push completed for {}", node);
                    lastBatchCount = status.getBatchesProcessed();
                } while (status.getReloadBatchesProcessed() > reloadBatchesProcessed && !status.failed());
            } finally {
                startTimesOfNodesBeingPushedTo.remove(node.getNodeId());
            }
        } else {
            log.warn("Cannot push to node '{}' in the group '{}'.  The sync url is blank",
                    node.getNodeId(), node.getNodeGroupId());
        }

    }

    private void pushToNode(Node remote, RemoteNodeStatus status) {
        Node identity = nodeService.findIdentity(false);
        NodeSecurity identitySecurity = nodeService.findNodeSecurity(identity.getNodeId());
        IOutgoingWithResponseTransport transport = null;
        ProcessInfo processInfo = statisticManager.newProcessInfo(new ProcessInfoKey(identity
                .getNodeId(), remote.getNodeId(), ProcessType.PUSH_JOB));
        try {
            transport = transportManager.getPushTransport(remote, identity,
                    identitySecurity.getNodePassword(), parameterService.getRegistrationUrl());

            List<OutgoingBatch> extractedBatches = dataExtractorService.extract(processInfo, remote, transport);
            if (extractedBatches.size() > 0) {
                
                log.info("Push data sent to {}", remote);
                
                List<BatchAck> batchAcks = readAcks(extractedBatches, transport, transportManager, acknowledgeService);
                status.updateOutgoingStatus(extractedBatches, batchAcks);
            }
            
            processInfo.setStatus(Status.DONE);
        } catch (Exception ex) {
            processInfo.setStatus(Status.ERROR);
            fireOffline(ex, remote, status);
        } finally {
            try {
                transport.close();
            } catch (Exception e) {
            }
        }
    }

}

File: symmetric-client-clib/src/service/PushService.c
/**
 * Licensed to JumpMind Inc under one or more contributor
 * license agreements.  See the NOTICE file distributed
 * with this work for additional information regarding
 * copyright ownership.  JumpMind Inc licenses this file
 * to you under the GNU General Public License, version 3.0 (GPLv3)
 * (the "License"); you may not use this file except in compliance
 * with the License.
 *
 * You should have received a copy of the GNU General Public License,
 * version 3.0 (GPLv3) along with this library; if not, see
 * <http://www.gnu.org/licenses/>.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "service/PushService.h"

SymList * SymPushService_readAcks(SymList *batches, SymOutgoingTransport *transport, SymTransportManager *transportManager, SymAcknowledgeService *acknowledgeService) {
    SymList *batchIds = SymList_new(NULL);
    SymIterator *iter = batches->iterator(batches);
    while (iter->hasNext(iter)) {
        SymOutgoingBatch *outgoingBatch = (SymOutgoingBatch *) iter->next(iter);
        if (SymStringUtils_equals(outgoingBatch->status, SYM_OUTGOING_BATCH_LOADING)) {
            batchIds->add(batchIds, SymStringUtils_format("%ld", outgoingBatch->batchId));
        }
    }
    iter->destroy(iter);

    SymList *batchAcks = transportManager->readAcknowledgement(transportManager, transport->ackString, transport->ackExtendedString);

    long batchIdInError = LONG_MAX;
    iter = batchAcks->iterator(batchAcks);
    while (iter->hasNext(iter)) {
        SymBatchAck *batchAck = (SymBatchAck *) iter->next(iter);
        char *batchId = SymStringUtils_format("%ld", batchAck->batchId);
        free(batchIds->removeObject(batchIds, batchId, (void *) strcmp));
        free(batchId);
        if (!batchAck->isOk) {
            batchIdInError = batchAck->batchId;
        }
        SymLog_debug("Saving ack: nodeId=%s, batchId=%ld, status=%s", batchAck->nodeId, batchAck->batchId,
                (batchAck->isOk ? SYM_OUTGOING_BATCH_OK : SYM_OUTGOING_BATCH_ERROR));
        acknowledgeService->ack(acknowledgeService, batchAck);
    }
    iter->destroy(iter);

    iter = batchIds->iterator(batchIds);
    while (iter->hasNext(iter)) {
        char *batchId = (char *) iter->next(iter);
        if (atol(batchId) < batchIdInError) {
            SymLog_error("We expected but did not receive an ack for batch %s", batchId);
        }
    }
    iter->destroy(iter);
    batchIds->destroyAll(batchIds, (void *) free);
    return batchAcks;
}

void SymPushService_pushToNode(SymPushService *this, SymNode *remote, SymRemoteNodeStatus *status) {
    SymNode *identity = this->nodeService->findIdentityWithCache(this->nodeService, 0);
    SymNodeSecurity *identitySecurity = this->nodeService->findNodeSecurity(this->nodeService, identity->nodeId);
    SymOutgoingTransport *transport = this->transportManager->getPushTransport(this->transportManager, remote, identity, identitySecurity->nodePassword,
            this->parameterService->getRegistrationUrl(this->parameterService));
    SymList *extractedBatches = this->dataExtractorService->extract(this->dataExtractorService, remote, transport);
    if (extractedBatches->size > 0) {
        SymLog_info("Push data sent to %s:%s:%s", remote->nodeGroupId, remote->externalId, remote->nodeId);
        SymList *batchAcks = this->readAcks(extractedBatches, transport, this->transportManager, this->acknowledgeService);
        status->updateOutgoingStatus(status, extractedBatches, batchAcks);
        batchAcks->destroy(batchAcks);
    }
    extractedBatches->destroy(extractedBatches);
    transport->destroy(transport);
    identitySecurity->destroy(identitySecurity);
}

void SymPushService_execute(SymPushService *this, SymNode *node, SymRemoteNodeStatus *status) {
    long reloadBatchesProcessed = 0;
    long lastBatchCount = 0;
    do {
        if (lastBatchCount > 0) {
            SymLog_info("Pushing to %s:%s:%s again because the last push contained reload batches",
                    node->nodeGroupId, node->externalId, node->nodeId);
        }
        reloadBatchesProcessed = status->reloadBatchesProcessed;
        SymLog_debug("Push requested for %s:%s:%s", node->nodeGroupId, node->externalId, node->nodeId);
        SymPushService_pushToNode(this, node, status);
        if (!status->failed && status->batchesProcessed > 0
                && status->batchesProcessed != lastBatchCount) {
            SymLog_info("Pushed data to %s:%s:%s. %ld data and %ld batches were processed",
                    node->nodeGroupId, node->externalId, node->nodeId, status->dataProcessed, status->batchesProcessed);
        } else if (status->failed) {
            SymLog_info("There was a failure while pushing data to %s:%s:%s. {} data and {} batches were processed",
                    node->nodeGroupId, node->externalId, node->nodeId, status->dataProcessed, status->batchesProcessed);
        }
        SymLog_debug("Push completed for %s:%s:%s", node->nodeGroupId, node->externalId, node->nodeId);
        lastBatchCount = status->batchesProcessed;
    } while (status->reloadBatchesProcessed > reloadBatchesProcessed && !status->failed);
}

SymRemoteNodeStatuses * SymPushService_pushData(SymPushService *this, unsigned int force) {
    SymMap *channels = this->configurationService->getChannels(this->configurationService, 0);
    SymRemoteNodeStatuses *statuses = SymRemoteNodeStatuses_new(NULL, channels);
    SymNode *identity = this->nodeService->findIdentity(this->nodeService);
    if (identity && identity->syncEnabled) {
        SymList *nodes = this->nodeCommunicationService->list(this->nodeCommunicationService, SYM_COMMUNICATION_TYPE_PUSH);
        if (nodes->size > 0) {
            SymNodeSecurity *identitySecurity = this->nodeService->findNodeSecurity(this->nodeService, identity->nodeId);
            if (identitySecurity) {
                SymIterator *iter = nodes->iterator(nodes);
                while (iter->hasNext(iter)) {
                    SymNode *node = (SymNode *) iter->next(iter);
                    SymRemoteNodeStatus *status = statuses->add(statuses, node->nodeId);
                    SymPushService_execute(this, node, status);
                }
                identitySecurity->destroy(identitySecurity);
                iter->destroy(iter);
            } else {
                SymLog_error("Could not find a node security row for '%s'.  A node needs a matching security row in both the local and remote nodes if it is going to authenticate to push data",
                        identity->nodeId);
            }
        }
        nodes->destroy(nodes);
    }
    channels->destroyAll(channels, (void *)SymChannel_destroy);
    return statuses;
}

void SymPushService_destroy(SymPushService *this) {
    free(this);
}

SymPushService * SymPushService_new(SymPushService *this, SymNodeService *nodeService, SymDataExtractorService *dataExtractorService,
        SymTransportManager *transportManager, SymParameterService *parameterService, SymConfigurationService *configurationService,
        SymAcknowledgeService *acknowledgeService, SymNodeCommunicationService *nodeCommunicationService) {
    if (this == NULL) {
        this = (SymPushService *) calloc(1, sizeof(SymPushService));
    }
    this->nodeService = nodeService;
    this->dataExtractorService = dataExtractorService;
    this->transportManager = transportManager;
    this->parameterService = parameterService;
    this->configurationService = configurationService;
    this->acknowledgeService = acknowledgeService;
    this->nodeCommunicationService = nodeCommunicationService;
    this->pushData = (void *) &SymPushService_pushData;
    this->readAcks = (void *) &SymPushService_readAcks;
    this->destroy = (void *) &SymPushService_destroy;
    return this;
}
